dhis-2/dhis-support/dhis-support-cache-invalidation/src/main/java/org/hisp/dhis/cacheinvalidation/EntityToDbTableMapping.java

```diff
@@ -0,0 +1,177 @@
+    {
+        if ( entityToTableNames.size() == 0 )
+        {
+            throw new IllegalStateException( "EntityToDbTableMapping is not initialized yet!" );
+        }
+        return entityToTableNames.get( tableName );
+    }
+
+    private void extractTableNamesFromHibernateMetamodel()
+    {
+        for ( EntityType<?> entity : sessionFactory.getMetamodel().getEntities() )
+        {
+            extractTableNames( entity.getJavaType() );
+        }
+    }
+
+    public void extractTableNames( final Class<?> modelClazz )
+    {
+        final MetamodelImpl metamodel = (MetamodelImpl) sessionFactory.getMetamodel();
+
+        final EntityPersister entityPersister = metamodel.entityPersister( modelClazz );
+        for ( int i = 0; i < entityPersister.getPropertyTypes().length; i++ )
+        {
+            Type type = entityPersister.getPropertyTypes()[i];
+            if ( type.isCollectionType() )
+            {
+                CollectionType collectionType = (CollectionType) type;
+                CollectionPersister collectionPersister = metamodel.collectionPersister( collectionType.getRole() );
+
+                if ( collectionPersister instanceof BasicCollectionPersister )
+                {
+                    BasicCollectionPersister bc = (BasicCollectionPersister) collectionPersister;
+                    String tableName = bc.getTableName();
+                    if ( entityToTableNames.get( tableName ) != null )
+                    {
+                        entityToTableNames.get( tableName )
+                            .add( new Object[] { modelClazz, collectionType.getRole() } );
+                    }
+                    else
+                    {
+                        List<Object[]> objects = new ArrayList<>();
+                        objects.add( new Object[] { modelClazz, collectionType.getRole() } );
+                        entityToTableNames.put( tableName, objects );
+                    }
+                }
+            }
+        }
+
+        if ( entityPersister instanceof SingleTableEntityPersister )
+        {
+            String tableName = ((SingleTableEntityPersister) entityPersister).getTableName();
+            if ( entityToTableNames.get( tableName ) != null )
+            {
+                entityToTableNames.get( tableName ).add( new Object[] { modelClazz } );
+            }
+            else
+            {
+                List<Object[]> objects = new ArrayList<>();
+                objects.add( new Object[] { modelClazz } );
+                entityToTableNames.put( tableName, objects );
+            }
+        }
+        else
+        {
+            throw new IllegalArgumentException( modelClazz + " does not map to a single table." );
+        }
+    }
+
+    private String printEntityTable( TreeMap<String, List<Object[]>> entityToTableNames )
+    {
+        StringBuilder sb = new StringBuilder();
+        Set<Map.Entry<String, List<Object[]>>> entries = entityToTableNames.entrySet();
```

Comment:
When types get extensive like this and you just want to loop the extra variable rather decreases readability IMHO.